modelmapper/loader.py

```diff
@@ -9,54 +10,131 @@ def select(fields):
+    """
+    Simple loader for inserting into a db with sqlalchemy.
+    """
+    def insert_row_into_db(self, row: dict, session, model):
+        """Insert row into db"""
+        row_obj = model(**row)
+        try:
+            session.add(row_obj)
+            session.flush()
+            return row_obj
+        except Exception:
+            self.logger.exception(f"Error on inserting row of data: {row}")
+            raise
+
+
+class SignatureSqlalchemyMixin(SqlalchemyLoaderMixin):
+    """
+    Base Signature loader. A signature column will be added to each row whose value is a 64-bit
+    murmur hash of the row (requiring BigInteger type). This will insert DUPLICATE ROWS!!!
+    If unique rows are desired, use the: UniqueSignatureSqlalchemyLoaderMixin.
+    """
 
-class PostgresSnapshotLoaderMixin():
+    def add_row_signature(self, chunk):
+        """Add hash of row to row about to be inserted"""
+        for row in chunk:
+            row['signature'] = generate_row_signature(
+                row, self.RECORDS_MODEL, self.settings.ignore_fields_in_signature_calculation
+            )
+            yield row
+
+    def get_id_by_signature(self, session, model, signature):
+        """Searches given table for given signature. Returns row if found or None if not"""
+        table = model.__table__
+        query = select([table.c.id]).where(table.c.signature == signature)
+        query_result = session.execute(query)
+        result = query_result.fetchone()
+        result = None if result is None else result[0]
+        return result
+
+    def insert_chunk_of_data_to_db(self, session, model, chunk):
+        """Add row signature to row then run Base class logic"""
+        new_chunk = self.add_row_signature(chunk)
+        return super().insert_chunk_of_data_to_db(session, model, new_chunk)
+
+
+class UniqueSignatureSqlalchemyLoaderMixin(SignatureSqlalchemyMixin):
+    """
+    Sqlalchemy Specific Loader with Signatures that DOES NOT ALLOW DUPLICATES.
+    It appends a signature (64-bit murmur hash of the row) and only uploads if
+    it is not already in the table.
     """
-    Postgres Specific Loader that backs up raw data into S3.
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.all_recent_rows_signatures = set()
+
+    def add_row_signature(self, chunk):
+        """Generate hash of row, yield row only if it has not been generated recently"""
+        chunk_gen = super().add_row_signature(chunk)
+        for row in chunk_gen:
+            signature = row.get('signature')
+            if signature and signature not in self.all_recent_rows_signatures:
+                self.all_recent_rows_signatures.add(signature)
+                yield row
+
+    def insert_row_into_db(self, row: dict, session, model):
+        """Insert row only if the signature is unique to the table"""
+        if row['signature'] and not self.get_id_by_signature(session, model, row['signature']):
```

Comment:
Why do we need to row by row check by signature manually instead of having Postgres throw an error when the signature is not unique?

Reply:
I will change this for the new loaders, have them rely on the table constraints